### PR TITLE
Ifc4x3 span annotation

### PR DESCRIPTION
**Pull request by seb-esser**
running discussion - needs improvements once it is clarified how to link the alignment and the annotation entity

### CODE REVIEW

**Comment by seb-esser** [comment]
@jmirtsch I've updated the test implementation but have some issues exporting `IfcCurveSegment` and `IfcAlignmentSegment`

File: ConsoleCreateSpanAnnotation/Program.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                0,
+                200,
+                IfcAlignmentHorizontalSegmentTypeEnum.LINE);
+
+            // geometric representation of a single segment. It gets referenced by IfcCompositeCurve.segments and IfcAlignmentSegment.Representation
+            var curveSegment = new IfcCurveSegment(
+                 IfcTransitionCode.CONTSAMEGRADIENTSAMECURVATURE,
+                 new IfcAxis2Placement2D(new IfcCartesianPoint(database, 0, 0)),
+                 new IfcParameterValue(0),
+                 new IfcParameterValue(200),
```
Comment (pjanck) [inline ConsoleCreateSpanAnnotation/Program.cs:57]: An `IfcLine` is parameterzied as [0,1]. If you want to give true length, use `IfcNonNegativeLengthMeasure`.

File: ConsoleCreateSpanAnnotation/Program.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                 null);
```
Comment (pjanck) [inline ConsoleCreateSpanAnnotation/Program.cs:58]: This is a non-optional attribute. You're missing a reference to `IfcLine` or similar.

File: ConsoleCreateSpanAnnotation/Program.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            alignmentSegment.Representation = new IfcProductDefinitionShape(new IfcShapeRepresentation(curveSegment));
+
+            #endregion Alignment
+
+            #region Annotation placement
+
+            alignmentSegment.AddComment("Create placement for annotation");
+
+            var axis2place = new IfcAxis2PlacementLinear(
+                new IfcPointByDistanceExpression(25, compositeCurve),
```
Comment (pjanck) [inline ConsoleCreateSpanAnnotation/Program.cs:98]: Does this 25 mean a ParameterValue or a NonNegativeLength?

File: ConsoleCreateSpanAnnotation/Program.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                null,
+                null);
```
Comment (pjanck) [inline ConsoleCreateSpanAnnotation/Program.cs:100]: Axis2PlacementLinear needs a reference to a curve along which it is defined. (`compositeCurve` probably in your case)

File: ConsoleCreateSpanAnnotation/Program.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+            var linPlacement = new IfcLinearPlacement(axis2place);
+            linPlacement.Distance = new IfcPointByDistanceExpression(128, compositeCurve);
```
Comment (pjanck) [inline ConsoleCreateSpanAnnotation/Program.cs:103]: no need to set this, deprecated

**Review by pjanck — CHANGES_REQUESTED** [review]
Some obligatory attributes left empty.

File: ConsoleCreateSpanAnnotation/Program.cs (hunk 3 above)
Comment (jmirtsch) [reply ConsoleCreateSpanAnnotation/Program.cs:98]: The constructor generates a NonNegativeLength if a double is provided, there's an overload that takes the measure value.

File: ConsoleCreateSpanAnnotation/Program.cs (hunk 2 above)
Comment (jmirtsch) [reply ConsoleCreateSpanAnnotation/Program.cs:58]: Yes, that's right.  Also note that now you can simply nominate the semantic segments for the horizontal, vertical and cant portions of an alignment into a constructor, and the geometric representation curve will be created for you (work in progress for some aspects such as transitions other than clothoids).

Look for this constructor, or perhaps we can have an online meeting and I help you adjust the code:
public IfcAlignmentHorizontal(IfcObjectPlacement placement, double startDistAlong, IEnumerable<IfcAlignmentHorizontalSegment> segments, out IfcCompositeCurve curveRepresentation)

**Pull request closed without merging**